packages/grid/x-data-grid-pro/src/hooks/features/columnResize/useGridColumnResize.tsx

```diff
@@ -441,10 +482,122 @@ export const useGridColumnResize = (
     { passive: doesSupportTouchActionNone() },
   );
 
+  useGridApiMethod(
+    apiRef,
+    {
+      autosizeColumns,
+    },
+    'public',
+  );
+
   useGridApiEventHandler(apiRef, 'columnSeparatorMouseDown', handleColumnResizeMouseDown);
   useGridApiEventHandler(apiRef, 'columnResizeStart', handleResizeStart);
   useGridApiEventHandler(apiRef, 'columnResizeStop', handleResizeStop);
 
   useGridApiOptionHandler(apiRef, 'columnResize', props.onColumnResize);
   useGridApiOptionHandler(apiRef, 'columnWidthChange', props.onColumnWidthChange);
 };
+
+function extractColumnWidths(
+  apiRef: React.MutableRefObject<GridPrivateApiPro>,
+  root: Element,
+  options: Parameters<GridColumnResizeApi['autosizeColumns']>[0],
+) {
+  type Result = Record<
+    string,
+    { min: number; max: number; contentMin: number; contentMax: number }
+  >;
+
+  const state = gridColumnsStateSelector(apiRef.current.state);
+
+  const inputColumns = options?.columns ?? state.orderedFields.map((field) => state.lookup[field]);
+  const columns = inputColumns.filter((c) => state.columnVisibilityModel[c.field] !== false);
+  const includeHeaders = options?.includeHeader ?? false;
+
+  try {
+    root.classList.add(gridClasses.autosizing);
+
+    const getHeader = (field: string) =>
+      root.querySelector(`[data-field="${field}"][role="columnheader"]`);
+
+    const getCells = (field: string) =>
+      Array.from(root.querySelectorAll(`[data-field="${field}"][role="cell"]`));
+
+    const widthsByField = columns.reduce((result, column) => {
+      const cells = getCells(column.field);
+      const widths = cells.map((cell) => {
+        // XXX:
+        // Is there a more efficient way to do this?
+        // Can we assume the cell padding is constant for all cells?
+        const style = window.getComputedStyle(cell, null);
+        const paddingWidth = parseInt(style.paddingLeft, 10) + parseInt(style.paddingRight, 10);
+        const contentWidth = cell.firstElementChild?.getBoundingClientRect().width ?? 0;
+        return paddingWidth + contentWidth;
+      });
```

Comment:
Not sure how to compute the width efficiently. I call `getComputedStyle` on every cell because I'm not sure if I can assume that the padding is constant. Comments welcome.

Reply:
I think this should be fine